tao_compiler/mlir/disc/transforms/disc_specialize_fusion_with_speculation.cc

```diff
@@ -479,17 +515,21 @@ struct DiscSpecializeFusionWithSpeculationPass
   }
 
   void runOnOperation() override {
-    // Stage #1: broadcast simplifier with speculation
+    // Stage #1: broadcast simplifier with speculation.
     Speculator(
         &DiscSpecializeFusionWithSpeculationPass::DoBroadcastSpeculation);
 
-    // Stage #2: speculation of reduce op
-    Speculator(
-        &DiscSpecializeFusionWithSpeculationPass::DoRowReductionSpeculation);
+    // Stage #2: speculation of column reduce op.
     Speculator(
         &DiscSpecializeFusionWithSpeculationPass::DoColReductionSpeculation);
 
-    // Stage #3: speculation of vectorization/tiling.
+    // Stage #3: speculation of row reduce op.
+    // We do row-reduction speculation before vectorization/tiling because TLP
+    // is usually more important than ILP on GPU.
+    Speculator(
+        &DiscSpecializeFusionWithSpeculationPass::DoRowReductionSpeculation);
+
```

Comment:
just curious, why do we swap the position of DoColReductionSpeculation & DoColReductionSpeculation?

Reply:
It actually does not matter. I used to try to merge the row-reduction and vec/tiling speculations into a holistic one. And I realized that the merging is not necessary. Thus leave row-reduce speculation just upon the one for vec/tiling. Moreover, the comments added for row-reduce is related to vec/tiling, and putting them together is slightly more readable.